saga/sagaState.go

```diff
@@ -0,0 +1,248 @@
+package saga
+
+import "errors"
+import "fmt"
+
+/*
+ * Data Structure representation of the current state of the Saga.
+ */
+type SagaState struct {
+	sagaId string
+	job    []byte
+
+	taskStarted   map[string]bool
+	taskCompleted map[string]bool
+	taskResults   map[string][]byte
+
+	compTaskStarted   map[string]bool
+	compTaskCompleted map[string]bool
+	compTaskResults   map[string][]byte
+
+	sagaAborted   bool
+	sagaCompleted bool
+}
+
+/*
+ * Initialize a Default Empty Saga
+ */
+func initializeSagaState() *SagaState {
+	return &SagaState{
+		sagaId:            "",
+		job:               nil,
+		taskStarted:       make(map[string]bool),
+		taskCompleted:     make(map[string]bool),
+		taskResults:       make(map[string][]byte),
+		compTaskStarted:   make(map[string]bool),
+		compTaskCompleted: make(map[string]bool),
+		compTaskResults:   make(map[string][]byte),
+		sagaAborted:       false,
+		sagaCompleted:     false,
+	}
+}
+
+/*
+ * Returns the Id of the Saga this state represents
+ */
+func (state *SagaState) SagaId() string {
+	return state.sagaId
+}
+
+/*
+ * Returns the Job associated with this Saga
+ */
+func (state *SagaState) Job() []byte {
+	return state.job
+}
+
+/*
+ * Returns true if the specified Task has been started,
+ * fasle otherwise
+ */
+func (state *SagaState) IsTaskStarted(taskId string) bool {
+	started, ok := state.taskStarted[taskId]
+	return started && ok
+}
+
+/*
+ * Returns true if the specified Task has been completed,
+ * fasle otherwise
+ */
+func (state *SagaState) IsTaskCompleted(taskId string) bool {
+	completed, ok := state.taskCompleted[taskId]
+	return completed && ok
+}
+
+/*
+ * Returns true if the specified Compensating Task has been started,
+ * fasle otherwise
+ */
+func (state *SagaState) IsCompTaskStarted(taskId string) bool {
+	started, ok := state.compTaskStarted[taskId]
+	return started && ok
+}
+
+/*
+ * Returns true if the specified Compensating Task has been completed,
+ * fasle otherwise
+ */
+func (state *SagaState) IsCompTaskCompleted(taskId string) bool {
+	completed, ok := state.compTaskCompleted[taskId]
+	return completed && ok
+}
+
+/*
+ * Returns true if this Saga has been Aborted, false otherwise
+ */
+func (state *SagaState) IsSagaAborted() bool {
+	return state.sagaAborted
+}
+
+/*
+ * Returns true if this Saga has been Completed, false otherwise
+ */
+func (state *SagaState) IsSagaCompleted() bool {
+	return state.sagaCompleted
+}
+
+/*
+ * Updates a SagaState with the specified SagaMessage.
+ * Returns an InvalidSagaState Error if applying the message would result in an invalid Saga State
+ * Returns an InvalidSagaMessage Error if the message is Invalid
+ */
+func (state *SagaState) UpdateSagaState(msg sagaMessage) error {
```

Comment:
I don't know, can do without the saga messages and factories? Are these messages truly opaque or do both SagaState and a real SagaLog impl have to look inside? If it's not opaque then I think it's cleaner to split UpdateSagaState() and LogMessage() into functions corresponding to each SagaMessage type.